lib/response-handler.js

```diff
@@ -18,53 +18,120 @@ module.exports = (bluebird, error, HttpLinkHeader, logger, ShelfError, URI) => {
 
 
     /**
-     * Maps error codes from the response to the internal error object.
-     * Returns the corresponding error.
+     * Creates a ShelfError from an error which contains a
+     * JSON response body with "code" and "message" properties.
+     * These are actual Shelf errors instead of things that
+     * could possibly come from a proxy or load balancer.
      *
-     * @param {(http~IncomingMessage|requestPromise~RequestError)} errorResponse
-     * @return {ShelfError}
+     * @see {@link https://visionmedia.github.io/superagent/#error-handling}
+     *
+     * @param {superagent~Error} reason
+     * @return {shelfLib~ShelfError}
      */
-    function createErrorForResponse(errorResponse) {
-        var errorCode, errorMessage;
+    function createErrorForShelf(reason) {
+        var body, errorCode, errorKey;
 
-        logger.debug("In responseHandler.handleErrorResponse()");
-        logger.debug(`Response param: ${JSON.stringify(errorResponse, null, 4)}`);
-
-        // Since the structure of error responses can be different based on
-        // where the error happens, we need to handle those different
-        // structures. This is handling when request-promise gives us a
-        // error reasonse.
-        if (errorResponse.error) {
-            errorResponse = errorResponse.response;
-        }
+        body = reason.response.body;
+
+        errorKey = Object.keys(error).find((key) => {
+            return error[key] === body.code;
+        });
 
-        // If the response has a code property, then it is a timeout error.
-        if (errorResponse.code) {
-            return new ShelfError("Socket Timeout", error.TIMEOUT);
+        if (errorKey) {
+            errorCode = error[errorKey];
+        } else {
+            errorCode = error.UNKNOWN;
         }
 
-        errorCode = Object.keys(error).find((key) => {
-            return error[key] === errorResponse.body.code;
+        return new ShelfError(body.message, errorCode);
+    }
+
+    /**
+     * Creates a ShelfError from any generic HTTP error.
+     * This only requires that a status property exists.
+     *
+     * @see {@link https://visionmedia.github.io/superagent/#error-handling}
+     *
+     * @param {superagent~Error} reason
+     * @return {shelfLib~ShelfError}
+     */
+    function createErrorForHttp(reason) {
+        var errorCode, errorCodeMap, mappedCode, statusCode;
+
+        errorCodeMap = {
+            404: error.NOT_FOUND,
+            500: error.INTERNAL_SERVER_ERROR
```

Comment:
Yea good question. Perhaps we could give the user something that might be helpful to us if someone posted it in the Shelf Helpline. I will leave that to you though, whatever you think works best. I think doing it the same way we are doing a 500 makes sense error.GATEWAY_TIMEOUT and error.SERVICE_UNAVAILABLE but I am not bother to much either way.